security-framework/src/trust_settings.rs

```diff
@@ -0,0 +1,253 @@
+
+                let maybe_name: Option<CFString> = settings.find(policy_name_key.as_CFTypeRef() as *const _)
+                    .map(|name| unsafe { CFString::wrap_under_get_rule(*name as *const _) });
+
+                match maybe_name {
+                    Some(ref name) if name != &ssl_policy_name => true,
+                    _ => false
+                }
+            };
+
+            if is_not_ssl_policy { continue; }
+
+            // Evaluate "effective trust settings" for this usage constraint.
+            let maybe_trust_result = {
+                let settings_result_key = CFString::from_static_string("kSecTrustSettingsResult");
+                settings.find(settings_result_key.as_CFTypeRef() as *const _)
+                    .map(|num| unsafe { CFNumber::wrap_under_get_rule(*num as *const _) })
+                    .and_then(|num| num.to_i64())
+            };
+
+            // "Note that an empty Trust Settings array means "always trust this cert,
+            //  with a resulting kSecTrustSettingsResult of kSecTrustSettingsResultTrustRoot"."
+            let trust_result = TrustSettingsForCertificate::new(maybe_trust_result
+                .unwrap_or(kSecTrustSettingsResultTrustRoot as i64));
+
+            match trust_result {
+                TrustSettingsForCertificate::Unspecified => { continue; },
+                TrustSettingsForCertificate::Invalid => { continue; },
+                _ => return Ok(trust_result),
+            }
+        }
+
+        // There were no more specific settings, and trust is the default.
+        Ok(TrustSettingsForCertificate::TrustRoot)
+    }
+}
+
+/// Iterator over certificates.
+pub struct TrustSettingsIter {
+    array: CFArray<SecCertificate>,
+    index: CFIndex,
+}
+
+impl Iterator for TrustSettingsIter {
+    type Item = SecCertificate;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.array.len() {
+            None
+        } else {
+            let cert = self.array.get(self.index)
+                .unwrap();
+            self.index += 1;
+            Some(cert.clone())
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let left = (self.array.len() as usize)
+            .saturating_sub(self.index as usize);
+        (left, Some(left))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn list_for_domain(domain: Domain) {
+        println!("--- domain: {:?}", domain);
+        let ts = TrustSettings::new(domain);
+        let iterator = ts.iter()
+            .unwrap();
+
+        for (i, cert) in iterator.enumerate() {
+            println!("cert({:?}) = {:?}", i, cert);
+            println!("  settings = {:?}", ts.tls_trust_settings_for_certificate(&cert));
+        }
+        println!("---");
+    }
+
+    #[test]
+    fn test_list_for_user() {
+        list_for_domain(Domain::User);
+    }
+
+    #[test]
+    fn test_list_for_system() {
+        list_for_domain(Domain::System);
```

Comment:
These aren't real tests. Could you check for some known cert? (some longer-lived one that macOS ships with. It's ok if the test breaks in a few years).

Reply:
Good idea. I've added one of these for the Let's Encrypt/ISRG root.